LLama/Native/NativeLogConfig.cs

```diff
@@ -0,0 +1,67 @@
+﻿using System.Runtime.InteropServices;
+using Microsoft.Extensions.Logging;
+
+namespace LLama.Native;
+
+/// <summary>
+/// Configure llama.cpp logging
+/// </summary>
+public class NativeLogConfig
+{
+    /// <summary>
+    /// Callback from llama.cpp with log messages
+    /// </summary>
+    /// <param name="level"></param>
+    /// <param name="message"></param>
+    public delegate void LLamaLogCallback(LLamaLogLevel level, string message);
+
+    /// <summary>
+    /// Register a callback to receive llama log messages
+    /// </summary>
+    /// <param name="logCallback"></param>
+    [DllImport(NativeApi.libraryName, CallingConvention = CallingConvention.Cdecl, EntryPoint = "llama_log_set")]
+    private static extern void native_llama_log_set(LLamaLogCallback? logCallback);
+
+    /// <summary>
+    /// Register a callback to receive llama log messages
+    /// </summary>
+    /// <param name="logCallback"></param>
+#pragma warning disable IDE1006 // Naming Styles (name imitates llama.cpp)
+    public static void llama_log_set(LLamaLogCallback? logCallback)
+#pragma warning restore IDE1006 // Naming Styles
+    {
+        if (NativeLibraryConfig.LibraryHasLoaded)
+        {
+            // The library is loaded, just pass the callback directly to llama.cpp
+            native_llama_log_set(logCallback);
+        }
+        else
+        {
+            // We can't set the log method yet since that would cause the llama.dll to load.
+            // Instead configure it to be set when the native library loading is done
+            NativeLibraryConfig.Instance.WithLogCallback(logCallback);
+        }
+    }
+
+    /// <summary>
+    /// Register a callback to receive llama log messages
+    /// </summary>
+    /// <param name="logger"></param>
+#pragma warning disable IDE1006 // Naming Styles (name imitates llama.cpp)
+    public static void llama_log_set(ILogger? logger)
+#pragma warning restore IDE1006 // Naming Styles
+    {
+        // Clear the logger
+        if (logger == null)
+        {
+            llama_log_set((LLamaLogCallback?)null);
+            return;
+        }
+
+        // Bind a function that converts into the correct log level
+        llama_log_set((level, message) =>
+        {
+            logger.Log(level.ToLogLevel(), "(llama.cpp): {message}", message);
+        });
```

Comment:
Not 100% sure, not an expert in P/Invoke, but probably the callback may be collected by the garbage collector (after the native call). I caught such exception when used an instance method as a callback (solved it by saving method as a delegate).
There should be no such problem with static methods/lambdas, but this callback captures logger, so it is not static.

Reply:
Added a GCHandle rooting the log callback when it's set.